### PR TITLE
Liquidations modeling streamlit

### PR DESCRIPTION
**Pull request by jepidoptera**

### CODE REVIEW

File: hydradx/model/indexer_utils.py [hunk 1]
```diff
@@ -40,12 +46,14 @@ def query_indexer(url: str, query: str, variables: dict = None) -> dict:
...
 
-    asset_query = """
-    query assetInfoByAssetIds($assetIds: [String!]!) {
-      assets(filter: {id: {in: $assetIds}}) {
-        nodes {
+    asset_query = f"""
+    query assetInfoByAssetIds{'($assetIds: [String!]!)' if asset_ids else ''} {{
+      assets(filter: {{{
+        'id: {in: $assetIds}, ' if asset_ids else ''
+      }symbol: {{notEqualTo: "none"}}}}) {{
```
Comment (poliwop) [inline hydradx/model/indexer_utils.py:55]: what's with the doubling up on curly braces?

File: hydradx/model/indexer_utils.py [hunk 2]
```diff
@@ -40,12 +46,14 @@ def query_indexer(url: str, query: str, variables: dict = None) -> dict:
     return return_val
 
 
-def get_asset_info_by_ids(asset_ids: list) -> dict:
+def get_asset_info_by_ids(asset_ids: list = None) -> dict:
```
Comment (poliwop) [inline hydradx/model/indexer_utils.py:49]: maybe add really basic test for this? Like, check that doing this query gives you a non-empty list of assets or something simple

File: hydradx/model/indexer_utils.py [hunk 3]
```diff
@@ -203,6 +192,24 @@ def get_omnipool_liquidity(
     return liquidity, hub_liquidity
 
 
+def get_current_block_height():
```
Comment (poliwop) [inline hydradx/model/indexer_utils.py:195]: I had not made this function yet because I'm not convinced it will be identical for every indexer we query. Maybe we should make the indexer an input? but then I'm not sure it'd be the same query for the different indexers?

When we are analyzing data from one of the storage indexers, we are probably going to want to know what the latest block number in the indexer in question was, not just the unified-prod indexer.

File: hydradx/model/indexer_utils.py [hunk 4]
```diff
@@ -390,6 +397,152 @@ def get_fee_pcts(data, asset_id):
...
+    for asset in asset_info.values():
+        if asset.asset_type == 'StableSwap':
+            asset.symbol = asset.name
+    current_block = get_current_block_height()
+    omnipool_data = get_omnipool_asset_data(min_block_id=current_block - 10000, max_block_id=current_block)
+    liquidity = {}
+    lrna = {}
+    for item in reversed(omnipool_data):
+        if asset_info[item['assetId']].symbol not in liquidity:
+            liquidity[asset_info[item['assetId']].symbol] = int(item['balances']['free']) / (10 ** asset_info[item['assetId']].decimals)
```
Comment (poliwop) [inline hydradx/model/indexer_utils.py:431]: isn't this going to add assets that were in Omnipool at some point in the last 10k blocks but are no longer in Omnipool? Is that what you are trying to do?

File: hydradx/model/indexer_utils.py [hunk 5]
```diff
@@ -390,6 +397,152 @@ def get_fee_pcts(data, asset_id):
...
+            lrna[asset_info[item['assetId']].symbol] = int(item['assetState']['hubReserve']) / (10 ** asset_info[1].decimals)
+        if len(liquidity) == len(asset_info):
```
Comment (poliwop) [inline hydradx/model/indexer_utils.py:433]: It looks like you think asset_info will only have omnipool assets, but in reality it should have many more assets than are in Omnipool, right?

File: hydradx/model/indexer_utils.py [hunk 6]
```diff
@@ -390,6 +397,152 @@ def get_fee_pcts(data, asset_id):
...
+            break
+    asset_fee, lrna_fee = get_current_omnipool_fees(asset_info)
+    for tkn in liquidity:
+        if tkn not in asset_fee.current:
+            asset_fee.current[tkn] = 0.0
```
Comment (poliwop) [inline hydradx/model/indexer_utils.py:438]: do we want to use default of 0 or min fees?

File: hydradx/model/indexer_utils.py [hunk 7]
```diff
@@ -390,6 +397,152 @@ def get_fee_pcts(data, asset_id):
...
+
+    if asset_info is None:
+        asset_info = get_asset_info_by_ids()
+    url = "https://galacticcouncil.squids.live/hydration-pools:unified-prod/api/graphql"
+    query = """
+        query MyQuery {
+            events(
+                first: 10000
+                orderBy: PARA_BLOCK_HEIGHT_DESC
+                filter: {name: {includes: "Omnipool"}}
```
Comment (poliwop) [inline hydradx/model/indexer_utils.py:466]: I think there are better ways to query for these fees... I would suggest looking in swaps table for trades involving "1" (hub asset) and looking at associated swap fees, or something similar. This doesn't have to change if it works but we shouldn't be needing to parse event args for this I think.

File: hydradx/tests/test_indexer_utils.py [hunk 8]
```diff
@@ -2,14 +2,16 @@
...
 from mpmath import mp, mpf
 
 import os
 os.chdir('../..')
 
-from hydradx.model.indexer_utils import get_latest_stableswap_data, get_stablepool_ids, get_fee_history, get_executed_trades, get_stableswap_liquidity_events
-
+from hydradx.model.indexer_utils import get_latest_stableswap_data, get_stablepool_ids, get_omnipool_liquidity, \
+    get_omnipool_asset_data, get_current_block_height, get_asset_info_by_ids, get_current_omnipool, \
+    get_current_omnipool_router, get_fee_history, get_fee_history, get_executed_trades, get_stableswap_liquidity_events
```
Comment (poliwop) [inline hydradx/tests/test_indexer_utils.py:14]: only need to import get_fee_history once

**Review by poliwop — CHANGES_REQUESTED** [review]
Some small changes to start with. Still need to review the changes to trade_strategies.py as well as the application logic itself.

File: hydradx/apps/money_market/toxic_debt.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+    evolve_function=update_prices
+)
+
+with st.spinner(f"Running {time_steps} simulation steps..."):
+    sim_start = time.time()
+    events = run(initial_state, time_steps=time_steps, silent=True)
+    sim_time = time.time() - sim_start
+    st.sidebar.info(f"Simulation completed in {sim_time:.2f} seconds")
+
+fig1, ax1 = plt.subplots()
```
Comment (poliwop) [inline hydradx/apps/money_market/toxic_debt.py:255]: I'm wondering if it's worth adding a check that the price collapse that the user configures does in fact happen? Maybe displaying a warning to the user when it doesn't happen for some reason?
What I mean is, e.g. in the default config where the DOT price collapses, we are graphing the DOT price in USD over time. I can visually look and see that seems to match the 75% crash. Let's maybe add a real check that this is the case, and a warning to the user if it fails for some reason?

File: hydradx/tests/test_money_market.py
Comment (poliwop) [inline hydradx/tests/test_money_market.py:1]: In test_liquidate_against_omnipool_no_liquidation(), agent needs to be using liquidate_cdps strategy.

**Review by poliwop — CHANGES_REQUESTED** [review]
Overall results look correct, some clarifications requested, some small refactoring and tests.

File: hydradx/model/amm/exchange.py [hunk 10]
```diff
@@ -20,16 +19,16 @@ def copy(self):
     def update(self):
         pass
 
-    def price(self, tkn: str, denomination: str = '') -> float:
+    def price(self, tkn: str, denominator: str = '') -> float:
```
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/model/amm/exchange.py]: changed all these to "numeraire"

File: hydradx/model/indexer_utils.py (hunk 1 above)
Comment (jepidoptera) [reply hydradx/model/indexer_utils.py:55]: It's for the formatting string. Double curly braces = literal curly brace, single = insert value of expression inside. I did this because in instances where you want all assets (not only certain ones) you can skip doing a query to get all asset ids and simplify this query by not including that filter.

File: hydradx/model/indexer_utils.py (hunk 4 above)
Comment (jepidoptera) [reply hydradx/model/indexer_utils.py:431]: You're right, and that would make the result incorrect for a day and a half or so. Do you know of a query to get only current assets?

File: hydradx/model/amm/money_market.py [hunk 11]
```diff
@@ -71,14 +74,17 @@ def __init__(
         self.emode_label = emode_label
 
 
-class MoneyMarket:
+class MoneyMarket(Exchange):
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/money_market.py:77]: In what way would it "force" the method to be implemented? (I'm just not familiar enough with abstract classes.) I would prefer it to not force anything unless someone tries to call the method, similar to duck typing?

But I think you can leave it as is for this PR, just something to think about for the future.

File: hydradx/model/indexer_utils.py (hunk 1 above)
Comment (poliwop) [reply hydradx/model/indexer_utils.py:55]: Interesting, I'm not sure I quite follow but I'll take a look. Sounds good, I was doing a whole separate query for that case (not filtering on asset ID).

File: hydradx/model/indexer_utils.py (hunk 2 above)
Comment (jepidoptera) [reply hydradx/model/indexer_utils.py:49]: done

File: hydradx/model/indexer_utils.py (hunk 4 above)
Comment (jepidoptera) [reply hydradx/model/indexer_utils.py]: ok, I added the current assets query and optimized so we don't have to look at all 10k blocks

File: hydradx/apps/money_market/toxic_debt.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+                tokens=tokens,
+                peg=peg,
+                amplification=pool.final_amplification,
+                trade_fee=pool.fee,
+                unique_id=pool_name,
+            )
+        )
+
+    mm = get_current_money_market()
+    # mm = MoneyMarket(
```
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/apps/money_market/toxic_debt.py:78]: I kept it there because if you're testing something, it saves 2-3 minutes of downloading the money market state (which is really slow).

File: hydradx/apps/money_market/toxic_debt.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+    usd_price_lrna = (
+        1 / load_omnipool.lrna_price('2-Pool-Stbl') / stable_swap_data[102].shares * 10 ** 18
+        * sum([int(v) / 10 ** asset_info[k].decimals for k, v in stable_swap_data[102].reserves.items()])
+    )
+    load_omnipool.add_token(
+        'USD', liquidity = usd_price_lrna, lrna=1
+    ).stablecoin = 'USD'
+    for pool in stable_swap_data.values():
+        pool_name = asset_info[pool.pool_id].name
+        if pool.pool_id == 690:
```
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/apps/money_market/toxic_debt.py:53]: What I'm doing is getting the gigadot pool shares from hydradxApi, getting the DOT/vDOT price from omnipool, then approximating the balances from that. I'll add a comment.

File: hydradx/model/amm/money_market.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/money_market.py:77]: I guess one way to accomplish what I have in mind would be to implement the dummy versions of these functions in the Exchange class to simply raise an exception if they ever get called.
Don't do this if it requires refactoring elsewhere, just a thought. The situation I am concerned about is where e.g. a price function returns 0 and we don't realize that's what's happened.

File: hydradx/tests/test_indexer_utils.py (hunk 8 above)
Comment (jepidoptera) [reply hydradx/tests/test_indexer_utils.py]: removed

File: hydradx/apps/money_market/toxic_debt.py [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+    print(f"Cache miss! Loading omnipool at {cache_time}")
+    load_router = get_current_omnipool_router()
+    load_omnipool = load_router.exchanges['omnipool']
+
+    asset_info = get_asset_info_by_ids()
+    stable_swap_data = get_stableswap_data()
+    stableswap_pools = []
+    usd_price_lrna = (
+        1 / load_omnipool.lrna_price('2-Pool-Stbl') / stable_swap_data[102].shares * 10 ** 18
+        * sum([int(v) / 10 ** asset_info[k].decimals for k, v in stable_swap_data[102].reserves.items()])
```
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/apps/money_market/toxic_debt.py:46]: added

File: hydradx/model/abi/pool_address_provider.py
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/model/abi/pool_address_provider.py:1]: Wait, actually the blank line was added, not deleted. So it's there already.

[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/abi/pool_address_provider.py:1]: confirmed, github diff's view is just very confusing

**Review by poliwop — APPROVED** [review]
Looks good
